src/digraph-cyclicity.jl

```diff
@@ -0,0 +1,371 @@
+    end
+    return dg
+end
+
+"""
+```ncycles_n_i(n::Integer, i::Integer)```
+
+Compute the theoretical maximum number of cycles of size `i` in a directed graph of `n`
+ vertices.
+"""
+function ncycles_n_i(n::Integer, i::Integer) 
+    return binomial(big(n), big(n-i+1)) * factorial(big(n-i))
+end
+
+"""
+```maxcycles(n::Integer)```
+
+Compute the theoretical maximum number of cycles in a directed graph of `n` vertices, 
+assuming there are no self-loops. 
+The formula is coming from [Johnson's paper](http://epubs.siam.org/doi/abs/10.1137/0204007).
+
+# Arguments
+* `n`: the number of vertices.
+"""
+
+function maxcycles(n::Integer)
+    return sum(x -> ncycles_n_i(n, x), 1:(n-1))
+end
+
+"""
+``` maxcycles(dg::DiGraph, byscc::Bool = true)```
+
+Compute the theoretical maximum number of cycles in the directed graph `dg`.
+
+The computation can be performed assuming the graph is complete or taking into account the 
+decomposition in strongly connected components. The formula is coming from 
+[Johnson's paper](http://epubs.siam.org/doi/abs/10.1137/0204007).
+
+# Arguments:
+* `dg`: The directed graph to be considered.
+* `byscc`: whether it should be computed knowing the strongly connected components of the
+ directed or not (default:yes)
+
+# Return
+* `c`: the theoretical maximum number of cycles.
+
+Note: A more efficient version is possible.
+"""    
+function maxcycles(dg::DiGraph, byscc::Bool = true) 
+    c = 0
+    n = dg.vertices[end]
+    if !byscc        
+        c = maxcycles(n)
+    else
+        for scc in strongly_connected_components(dg)
+            length(scc) > 1 && (c += maxcycles(length(scc)))
+        end     
+    end
+    return c
+end
+
+"""
+```type JohnsonVisitor <: Visitor
+    stack::Vector{Int}
+    blocked::Dict{Int, Bool}
+    blockedmap::Dict{Int, Set{Int}}
+end```
+
+Composite type that regroups the information needed for Johnson's algorithm.
+
+# Arguments
+* `stack`: the stack of visited vertices
+* `blocked`: boolean for each vertex, tell whether it is blocked or not
+* `blockedmap`: tell which vertices to unblock if the key vertex is unblocked
+"""
+type JohnsonVisitor <: Visitor
+    stack::Vector{Int}
+    blocked::Vector{Bool}
+    blockedmap::Vector{Set{Int}}
+end
+
+"""
+```JohnsonVisitor(dg::DiGraph)```
+
+Constructor of the visitor, using the directed graph information.
+"""
+JohnsonVisitor(dg::DiGraph) = JohnsonVisitor(
+Vector{Int}(),
+falses(dg.vertices),
+[Set{Int}() for i in dg.vertices]
+)
+
+"""
+```unblock!(v::T, blocked::Vector{Bool}, B::Vector{Set{Int}})```
+
+Unblock the vertices recursively.
+
+# Arguments
+* `v`: the vertex to unblock
+* `blocked`: tell whether a vertex is blocked or not
+* `B`: the map that tells if the unblocking of one vertex should unblock other vertices
+"""
+function unblock!(v::Int, blocked::Vector{Bool}, B::Vector{Set{Int}})
+    blocked[v] = false
+    for w in B[v]
+        delete!(B[v], w)
+        blocked[w] && unblock!(w, blocked, B)
+    end
+end
```

Comment:
I am not sure of this. Probably as the NetworkX version of this algorithm is done iteratively and not recursively.
The large stack problem is more in the circuit recursion than in the unblock one though.